TRITON-2028 - Want cmon-agent to export zpool status by default

khalfella opened the pull request.

KodyKantor commented on lib/instrumenter/collectors-gz/zpool.js:
```diff
@@ -70,6 +70,28 @@ function getZpoolKstats(opts, callback) {
+ * Encode zpool health status into a numberic value.
+ */
+function zpoolHealthEnc(health) {
+    var zpoolStatus = {
```
We could make this an array and do something like
```
return zpoolStatus.indexOf(health);
```
That should behave the same way as this code and be a little more concise.

khalfella replied:
Nice idea - The code has been updated to use `indexOf()`.

KodyKantor commented on lib/instrumenter/collectors-gz/zpool.js:
```diff
@@ -134,6 +151,11 @@ function getZpoolStats(callback) {
                 metric.value = z[ind].trim();
+
+                if (keys[ind].encoder) {
+                    metric.value = keys[ind].encoder(metric.value);
```
I just have a couple questions here:
- what if the encoder returns -1 like the error case for `zpoolHealthEnc`? Maybe we can assert that all encoders return >= 0, and a negative return value means something undefined happened? Or is it valuable to return -1 in the formatted metrics?
- I'm guessing we don't need to run `trim()` on these since we're relying on another function to give us the string to use. Maybe we can wrap the `z[ind].trim()` line in an `else` block to make that clear.

khalfella replied:
- Yes, `-1` means `zpoolHealthEnc()` failed to encode pool health status. This should happen if ZFS code is updated and a new pool status is introduced. I updated the code to print a warning message if that happens. However, that doesn't mean we should abort the agent or stop processing this metric. IMO, `-1` tells the user that the pool is not online and we don't know its current status. I think returning `-1` is better than returning no metric.

- Assuming `zpool list -Hp` output doesn't contain spaces, I don't think we don't need `z[ind].trim()`, I removed it from the code. I don't know why it was there in the first place? I don't think we need an `else` block. We should be able to use `zpool list` output as it is for metrics with no encoders.

KodyKantor commented:
Great, looks good. It would be great if we could somehow document to cmon-agent consumers what the numbers correspond to, but I'm not sure there's a way to do that aside from writing it all in the `# HELP` text.

Were you able to run the cmon-agent test suite? There might be a test or two that need updating after this change.

khalfella commented:
@KodyKantor  - Apologies for taking all this time to get back to this change. I updated the description of the metrics to show what the pool status numbers correspond to. Now the help message looks better. It is helpful now.

```
# HELP zpool_health_status the current health of the pool (0 = ONLINE, 1 = DEGRADED, 2 = FAULTED, 3 = OFFLINE, 4 = REMOVED, 5 = UNAVAIL, -1 = UNKNOWN)
# TYPE zpool_health_status gauge
zpool_health_status{pool="zones"} 0
```
I also added tests for zpool metrics.

KodyKantor commented on test/zpool.test.js:
```diff
@@ -0,0 +1,96 @@
+            return;
+        }
+
+        var valid_metric_types = ['counter', 'gauge', 'histogram'];
```
It looks like the zpool metric code only uses 'gauge' type metrics at this point. Should we only include 'gauge' in this list? That might help to make sure we don't introduce backward-incompatible changes down the line.

khalfella replied:
Good catch. We have two `counter` metrics `metaslab_group_loads` and `metaslab_group_unloads`, but we don't have histograms. I removed `histogram` from the list.

KodyKantor replied:
Great, thanks for clarifying that.

khalfella replied:
Thank you for taking the time to review this work.
